### PR TITLE
MULE-19208: HTTP Service response handling threads get stuck when an …

### PR DESCRIPTION
**Pull request by eze210**
…error is detected before the first chunk

### CODE REVIEW

**Comment by eze210** [comment]
--validate

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 1]
```diff
@@ -254,4 +281,98 @@ private void handleIfNecessary() {
       }
     }
   }
+
+  /**
+   * Decorator used to avoid selectors from being blocked reading from not yet written streams.
+   *
```
Comment (fsgonz) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java:287]: empty line

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 2]
```diff
@@ -254,4 +281,98 @@ private void handleIfNecessary() {
...
+      }
+      return super.read(b);
+    }
+
+    /**
+     * Same as parent except that it returns <code>0</code> if nobody wrote in the other side of the stream yet and the
+     * stream is not closed.
+     *
+     * @param b Buffer to fill by the method.
+     * @return The number of bytes read if available;
```
Comment (fsgonz) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java:320]: remove semicolon

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 3]
```diff
@@ -254,4 +281,98 @@ private void handleIfNecessary() {
...
+     *         <code>-1</code> if the stream is closed;
```
Comment (fsgonz) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java:321]: remove semicolon

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 4]
```diff
@@ -254,4 +281,98 @@ private void handleIfNecessary() {
...
+    DecoratedPipedInputStream(PipedOutputStream output, int bufferSize) throws IOException {
+      super(output, bufferSize);
+    }
+
+    /**
+     * Same as parent except that it returns <code>0</code> if nobody wrote in the other side of the stream yet and the
+     * stream is not closed.
+     *
+     * @param b Buffer to fill by the method.
+     * @return The number of bytes read if available;
```
Comment (fsgonz) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java:303]: remove semicolon

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 5]
```diff
@@ -254,4 +281,98 @@ private void handleIfNecessary() {
...
+
+  private class DecoratedPipedOutputStream extends PipedOutputStream {
+
+    private DecoratedPipedInputStream countingSink;
+
+    @Override
+    public synchronized void connect(PipedInputStream snk) throws IOException {
+      super.connect(snk);
+      if (!(snk instanceof DecoratedPipedInputStream)) {
+        throw new IllegalArgumentException("Sink must be an instance of CountingPipedInputStream");
```
Comment (fsgonz) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java:349]: shouldn't this be DecoratedPipedInputStream?

**Review by fsgonz — CHANGES_REQUESTED** [review]

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java (hunk 5 above)
Comment (eze210) [reply src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java:349]: Obviously!

**Review by fsgonz — APPROVED** [review]

**Review by fsgonz — APPROVED** [review]

**Pull request merged**